mender-update/inventory.cpp

```diff
@@ -0,0 +1,171 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <mender-update/inventory.hpp>
+
+#include <sstream>
+#include <string>
+
+#include <common/api.hpp>
+#include <common/common.hpp>
+#include <common/error.hpp>
+#include <common/events.hpp>
+#include <common/http.hpp>
+#include <common/inventory_parser.hpp>
+#include <common/io.hpp>
+#include <common/json.hpp>
+#include <common/log.hpp>
+#include <common/path.hpp>
+
+namespace mender {
+namespace update {
+namespace inventory {
+
+using namespace std;
+
+namespace api = mender::common::api;
+namespace common = mender::common;
+namespace error = mender::common::error;
+namespace events = mender::common::events;
+namespace expected = mender::common::expected;
+namespace http = mender::http;
+namespace inv_parser = mender::common::inventory_parser;
+namespace io = mender::common::io;
+namespace json = mender::common::json;
+namespace log = mender::common::log;
+namespace path = mender::common::path;
+
+const InventoryErrorCategoryClass InventoryErrorCategory;
+
+const char *InventoryErrorCategoryClass::name() const noexcept {
+	return "InventoryErrorCategory";
+}
+
+string InventoryErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	case BadResponseError:
+		return "Bad response error";
+	}
+	assert(false);
+	return "Unknown";
+}
+
+error::Error MakeError(InventoryErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, InventoryErrorCategory), msg);
+}
+
+const string uri = "api/devices/v1/inventory/device/attributes";
+
+error::Error PushInventoryData(
+	const string &inventory_generators_dir,
+	const string &server_url,
+	http::Client &client,
+	events::EventLoop &loop,
+	APIResponseHandler api_handler) {
+	auto ex_inv_data = inv_parser::GetInventoryData(inventory_generators_dir);
+	if (!ex_inv_data) {
+		return ex_inv_data.error();
+	}
+
+	stringstream top_ss;
+	top_ss << "[";
```

Comment:
S: No need to do this manually, we have functionality for this:

```
diff --git a/mender-update/inventory.cpp b/mender-update/inventory.cpp
index fd34ba82..bdf9d583 100644
--- a/mender-update/inventory.cpp
+++ b/mender-update/inventory.cpp
@@ -14,7 +14,6 @@
 
 #include <mender-update/inventory.hpp>
 
-#include <sstream>
 #include <string>
 
 #include <common/api.hpp>
@@ -80,31 +79,11 @@ error::Error PushInventoryData(
 		return ex_inv_data.error();
 	}
 
-	stringstream top_ss;
-	top_ss << "[";
-	auto inv_data = ex_inv_data.value();
-	for (const auto &kv : inv_data) {
-		top_ss << R"({"name":")";
-		top_ss << json::EscapeString(kv.first);
-		top_ss << R"(","value":)";
-		if (kv.second.size() == 1) {
-			top_ss << "\"" + json::EscapeString(kv.second[0]) + "\"";
-		} else {
-			stringstream items_ss;
-			items_ss << "[";
-			for (const auto &str : kv.second) {
-				items_ss << "\"" + json::EscapeString(str) + "\",";
-			}
-			auto items_str = items_ss.str();
-			// replace the trailing comma with the closing square bracket
-			items_str[items_str.size() - 1] = ']';
-			top_ss << items_str;
-		}
-		top_ss << R"(},)";
+	auto exp_payload {json::Dump(ex_inv_data.value())};
+	if (!exp_payload) {
+		return exp_payload.error();
 	}
-	auto payload = top_ss.str();
-	// replace the trailing comma with the closing square bracket
-	payload[payload.size() - 1] = ']';
+	auto payload = exp_payload.value();
 
 	http::BodyGenerator payload_gen = [payload]() {
 		return make_shared<io::StringReader>(payload);
```

Reply:
That doesn't convert one-item vectors to plain string values, right?

Reply:
> That doesn't convert one-item vectors to plain string values, right?

Good point, this is true!

> Plus it doesn't add the "name" and "value" things. The input key-value data doesn't have anything like that

Nope, it doesn't, and yeh, maybe ur right. That this is the way :sigh: